pageserver/src/utilization.rs

```diff
@@ -34,13 +39,31 @@ pub(crate) fn regenerate(tenants_path: &Path) -> anyhow::Result<PageserverUtiliz
 
     let captured_at = std::time::SystemTime::now();
 
+    let (disk_wanted_bytes, shard_count) = tenant_manager.calculate_utilization()?;
+
+    // Calculate how much disk space can actually be used (we will not use disk space
+    // beyond the eviction threshold)
+    let max_usage_pct = match conf.disk_usage_based_eviction.clone() {
+        Some(e) => e.max_usage_pct,
+        None => Percent::new(100).unwrap(),
+    };
+    let disk_usable_capacity = ((used + free) * max_usage_pct.get() as u64) / 100;
+
+    // Utilization score will be a number scaled to 100.  The scale is totally arbitrary: these
+    // scores are only meaningful when compared to one another.  The value can go over 100
+    // if we are uncomfortably overloaded.
+    let disk_utilization_score = disk_wanted_bytes * 100 / disk_usable_capacity;
+
+    // In case we have many tiny tenants, also apply an absolute limit on the number of
+    // shards we will host on one pageserver
+    const MAX_SHARDS: usize = 20000;
+    let shard_utilization_score = (shard_count * 100 / MAX_SHARDS) as u64;
```

Comment:
Why overload the utilisation score? The storage controller already knows how many shards are present on each pageserver, so it could do this inference on its own.

Reply:
It's not essential.  The higher level question would be: why have a utilisation score at all, rather than just giving the controller the inputs?  Our original answer to that was that it would be used by the control plane, but this seems less likely now.

So we could kind of do it either way...

I kind of like the idea of the pageserver owning this calculation though: it means we can evolve it to include more things in future (e.g. if a node is heavily CPU-loaded) without having to thread the underlying telemetry up into the controller.

Reply:
Thinking ahead to using this in the controller: we'll want the ability to adjust a score based on scheduling changes we make.  So it probably makes sense to write PageserverUtilization as a type that carries all the inputs, can be modified on the controller, and has a score() method for giving the actual scheduling score.

Reply:
Re-worked this so that PageserverUtilization treats `utilization_score` as just a cache of its `score()` function's output.  It's kind of redundant to serialize it, but I'm leaving it in because I expect we'll calculate it anyway when we add a metric for this, and it makes the API easier for a human to eyeball